### PR TITLE
Added overload distortion effect, used by a few tracks in CBFD

### PR DESCRIPTION
**Pull request by derselbst**
This implements a missing audio effect found in CBFD. It can be heard in sparse41, sparse43, and possibly elsewhere. With this change, we are finally able to hear Surf Punks in its full HLE glory. Take a listen:

[sparse41.zip](https://github.com/mupen64plus/mupen64plus-rsp-hle/files/8063193/sparse41.zip)

I haven't checked whether this effect is also present in JFG.

All credits go to @joshware.

### CODE REVIEW

**Comment by derselbst** [comment]
@bsmiles32 FYI

**Comment by joshware** [comment]
Hey @derselbst 

I've taken a look at microcode for Jet Force Gemini, and the same code is present there.

**Comment by derselbst** [comment]
Interesting. Perhaps we should apply this change for `alist_process_naudio_mp3` as well. But then the question remains how `naudio_mp3` differs from `naudio_cbfd`.

Perhaps a bit off-topic: How exactly are you looking at the micrcode? Are you using the debug output of the Pure Interpreter? Or are you using IDA? And how do you isolate the right mircocode section?

**Comment by joshware** [comment]
I'm dumping the RSP memory when it changes, and then loading it into IDA.

I've looked at all the known `naudio_mp3` games, and they all have the overload function.
Which indeed does beg the question of what is the difference between it and the `naudio_cbfd` code.

So, I've tried to side-by-side them, and it appears that despite being a newer game, CBFD appears to have a slightly older version compared to JFG, B.T. et al.

For `naudio_mp3`, the functions DMEM parameters have an additional protective AND on them (basically dmem & 0xffff).
But there are minor differences are in the RESAMPLE and ENVMIXER functions.

I don't think it is making any noticeable difference, as it could be just a simplification of the logic.
And the code in alist.c doesn't appear to be a exact implantation to either of them.

tl;dr: They aren't the same, but merging them wouldn't break CBFD, and the rest would use the new overload function - if the games use it at all.

**Comment by derselbst** [comment]
> CBFD appears to have a slightly older version compared to JFG, B.T. et al.

Interesting. Maybe the developers spent their spare time with RSP programming when TTC64 was redesigned into CBFD.

Ok, let's wait for @bsmiles32, maybe he has an opinion about how to deal with `naudio_cbfd` vs `naudio_mp3`.

File: src/alist.c [hunk 1]
```diff
@@ -1022,3 +1022,17 @@ void alist_iirf(
     dram_store_u16(hle, (uint16_t*)&ibuf[(index-1)&3], address+10, 1);
 }
 
+/* Perform a clamped gain, then attenuate it back by an amount */
+void alist_overload(struct hle_t* hle, uint16_t dmem, int16_t count, int16_t gain, int16_t attenuation)
```
Comment (bsmiles32) [inline src/alist.c:1026]:
```suggestion
void alist_overload(struct hle_t* hle, uint16_t dmem, int16_t count, int16_t gain, uint16_t attenuation)
```

The ucode uses vmudh (s16 x s16) then vmudm (s16 x u16), so I think that attenuation should be unsigned instead.

File: src/alist.h [hunk 2]
```diff
@@ -150,6 +150,13 @@ void alist_iirf(
         int16_t* table,
         uint32_t address);
 
+void alist_overload(
+        struct hle_t* hle,
+        uint16_t dmem,
+        int16_t count,
+        int16_t gain,
+        int16_t attenuation);
```
Comment (bsmiles32) [inline src/alist.h:158]:
```suggestion
        uint16_t attenuation);
```

File: src/alist_naudio.c [hunk 3]
```diff
@@ -263,6 +263,16 @@ static void MP3(struct hle_t* hle, uint32_t w1, uint32_t w2)
     mp3_task(hle, index, address);
 }
 
+static void OVERLOAD(struct hle_t* hle, uint32_t w1, uint32_t w2)
+{
+    /* Overload distortion effect for Conker's Bad Fur Day */
+    uint16_t dmem = (w1 & 0xfff) + NAUDIO_MAIN;
+    int16_t gain = w2 & 0x7fff;
+    int16_t attenuation = w2 >> 16;
```
Comment (bsmiles32) [inline src/alist_naudio.c:271]:
```suggestion
    uint16_t attenuation = w2 >> 16;
```

File: src/alist_naudio.c [hunk 4]
```diff
@@ -263,6 +263,16 @@ static void MP3(struct hle_t* hle, uint32_t w1, uint32_t w2)
     mp3_task(hle, index, address);
 }
 
+static void OVERLOAD(struct hle_t* hle, uint32_t w1, uint32_t w2)
+{
+    /* Overload distortion effect for Conker's Bad Fur Day */
+    uint16_t dmem = (w1 & 0xfff) + NAUDIO_MAIN;
+    int16_t gain = w2 & 0x7fff;
```
Comment (bsmiles32) [inline src/alist_naudio.c:270]:
```suggestion
    int16_t gain = (int16_t)(uint16_t)w2;
```

The ucode uses mtc2 to put the gain in vector register, there is no masking taking place as far as I can see (but I may have missed something, it's been a long time I haven't touched any RSP assembly...)

File: src/alist_naudio.c [hunk 5]
```diff
@@ -320,9 +330,8 @@ void alist_process_naudio_mp3(struct hle_t* hle)
 
 void alist_process_naudio_cbfd(struct hle_t* hle)
 {
-    /* TODO: see what differs from alist_process_naudio_mp3 */
     static const acmd_callback_t ABI[0x10] = {
-        UNKNOWN,        ADPCM,          CLEARBUFF,      ENVMIXER,
+        OVERLOAD,       ADPCM,          CLEARBUFF,      ENVMIXER,
```
Comment (bsmiles32) [inline src/alist_naudio.c:334]: You might also want to add OVERLOAD to alist_process_naudio_mp3 as it also uses this very same audio command (no changes).

The only difference I could remember between mp3 and cbfd variants is in the MP3ADDY command, which is btw not implemented in rsp-hle and whose name is not really meaningful.

Still I would prefer if both variants are kept separated for documentation purpose.

**Review by bsmiles32 — CHANGES_REQUESTED** [review]
Sorry it took so long to review your PR.
Overall it's good work ! I've added some comments where I feel we can improve your implementation.
Thanks for your contribution

File: src/alist_naudio.c (hunk 5 above)
Comment (derselbst) [reply src/alist_naudio.c:346]: Thanks for the review! I've tested the unsigned version and it works as well (little surprising). I also added a comment containing the finding of the ucode differences.

**Comment by derselbst** [comment]
@kode54 FYI: You might want to include this change in foo_input_usf / lazyusf2.

**Comment by mudlord** [comment]
I might do the same for libretro mupen builds too, considering its now official.
